MSAL/src/MSALPublicClientApplication.m

```diff
@@ -174,140 +202,124 @@ - (id)initWithClientId:(NSString *)clientId
-    {
-        if (error) *error = [MSALErrorConverter msalErrorFromMsidError:msidError];
-        return nil;
-    }
-
-    MSALRedirectUri *msalRedirectUri = [MSALRedirectUriVerifier msalRedirectUriWithCustomUri:redirectUri
-                                                                                    clientId:clientId
+    MSALRedirectUri *msalRedirectUri = [MSALRedirectUriVerifier msalRedirectUriWithCustomUri:config.redirecrUri
+                                                                                    clientId:config.clientId
                                                                                        error:&msidError];
-
+    
     if (!msalRedirectUri)
     {
         if (error) *error = [MSALErrorConverter msalErrorFromMsidError:msidError];
         return nil;
     }
-
-    _redirectUri = msalRedirectUri;
-
+    
+    config.verifiedRedirectUri = msalRedirectUri;
+    
 #if TARGET_OS_IPHONE
     // Optional Paramater
-    _keychainGroup = keychainGroup;
-
-    MSIDKeychainTokenCache *dataSource;
-    if (_keychainGroup == nil)
-    {
-        _keychainGroup = [[NSBundle mainBundle] bundleIdentifier];
-    }
+    MSIDKeychainTokenCache *dataSource = [[MSIDKeychainTokenCache alloc] initWithGroup:MSALGlobalConfig.cacheConfig.keychainSharingGroup];
     
-    dataSource = [[MSIDKeychainTokenCache alloc] initWithGroup:_keychainGroup];
-
     MSIDLegacyTokenCacheAccessor *legacyAccessor = [[MSIDLegacyTokenCacheAccessor alloc] initWithDataSource:dataSource otherCacheAccessors:nil];
     MSIDDefaultTokenCacheAccessor *defaultAccessor = [[MSIDDefaultTokenCacheAccessor alloc] initWithDataSource:dataSource otherCacheAccessors:@[legacyAccessor]];
     
     self.tokenCache = defaultAccessor;
-    
-    _webviewType = MSALWebviewTypeDefault;
-    
 #else
     __auto_type dataSource = MSIDMacTokenCache.defaultCache;
-
+    
     MSIDDefaultTokenCacheAccessor *defaultAccessor = [[MSIDDefaultTokenCacheAccessor alloc] initWithDataSource:dataSource otherCacheAccessors:nil];
     self.tokenCache = defaultAccessor;
-    _webviewType = MSALWebviewTypeWKWebView;
-    
 #endif
-    
-    _validateAuthority = YES;
-    _extendedLifetimeEnabled = NO;
-    
-    _sliceParameters = [MSALPublicClientApplication defaultSliceParameters];
+    _configuration = config;
     
     MSIDAADNetworkConfiguration.defaultConfiguration.aadApiVersion = @"v2.0";
-    _expirationBuffer = 300;  //in seconds, ensures catching of clock differences between the server and the device
     
     return self;
 }
 
+- (id)initWithClientId:(NSString *)clientId
+         keychainGroup:(NSString *)keychainGroup
+             authority:(MSALAuthority *)authority
+           redirectUri:(NSString *)redirectUri
+                 error:(NSError * __autoreleasing *)error
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+
+#if TARGET_OS_IPHONE
+    if (keychainGroup) MSALGlobalConfig.cacheConfig.keychainSharingGroup = keychainGroup;
```

Comment:
previous logic was always passing defaultKeychainGroup to acquire token calls when developer didn't specify one. New logic is "self.keychainGroup" to this call. The question is how will developer get the default behavior?

Reply:
Correct, self.keychainGroup by default will always pass what used to be the default keychain unless it's changed.

Reply:
This should be same as before.

1. keychainGroup is specified with nil value: [[NSBundle mainBundle] bundleIdentifier]
2. keychainGroup is specified with non-nil value, custom value is used.
3. keychainGroup is not specified: use default - com.microsoft.adalcache

3 happens for all initializers that does not have keychainGroup as a parameter.
1 and 2 happens for the ones that have keychainGroup as a parameter, in which dev can opt to pass nil or a value.